### PR TITLE
makepy: support for enums and type-hints

### PR DESCRIPTION
**Pull request by tobias-loew**
Hi,
I've added command flags to generate enum and type-hints.
I've tested it with a quite large typelib (generated py-file has > 100k lines) and it seems to work.
Regarding the enums: there's also code for creating "Enums" (instead of "IntEnums"; use createEnums == 1) but a could find a way to make properties working without either putting a runtime penalty on everyone or breaking the ABI in __init__.py. Maybe you've got an idea!
Tobias

### CODE REVIEW

**Comment by thopiekar** [comment]
Hello @tobias-loew !
Thank you for your contribution. :+1: 

Can you tell us which typelib has been used to test your work? Work be great to know to be able to reproduce your work. 
Additionally, I still see some commented out lines in your code. Is there still some work to do to finish your PR? Are these to-dos you want to fix in the future?

Thanks! 🙂

**Comment by tobias-loew** [comment]
Hi,
sorry for the commented code - I'll have a look at it.
I tested the extension with EbsOpen.tlb, the COM-Library of the product I'm the lead developer (EBSILONProfessional, www.ebsilon.com)
[EbsOpen.zip](https://github.com/mhammond/pywin32/files/3416132/EbsOpen.zip)

**Comment by tobias-loew** [comment]
I've import the "Microsoft Scripting Runtime" three times: the old way, with enums and with enums & typehints.
[ScriptingRuntime.zip](https://github.com/mhammond/pywin32/files/3599577/ScriptingRuntime.zip)

And here is the imported "with eunms and typehints" version of EbsOpen.py (the product I develop)
[EbsOpen.zip](https://github.com/mhammond/pywin32/files/3599585/EbsOpen.zip)

**Comment by tobias-loew** [comment]
The extension for enums and type-hints require Python 3. So, all changes here target only Python 3.
I'm not a Python expert, so I don't know how to (if possible) hide code from Python 2.

**Comment by CarliJoy** [comment]
@tobias-loew to hide code from Python 2 use:
```
import sys
if sys.version_info >= (3, 0):
  your_python_3_code()
```

@mhammond given that version 300 was just released, the problems with 2to3 changes should be resolved or?

**Comment by mhammond** [comment]
> @mhammond given that version 300 was just released, the problems with 2to3 changes should be resolved or?

Yes, please!

**Comment by tobias-loew** [comment]
I've resolved the conflicts but didn't yet have time to make the "sys.version_info >= (3, 0)" changes.

**Comment by tobias-loew** [comment]
IMHO the best way is to add an additional option to support the generation of files with enums for version 3 AND the old style for version 2.
Thus, python 3- only users won't have to suffer from the runtime overhead.
Any objections?

**Comment by CarliJoy** [comment]
Python 2 is not supported anymore. You can drop any Python 2 code.
Also pywin32 droppped Python 2 support beginning with Version 300. So you can simply ignore Python 2.
We just have to find out why test test are failing (currently Github has troubles downloading the log for me)

**Comment by kxrob** [comment]
> We just have to find out why test test are failing (currently Github has troubles downloading the log for me)

Thats because of the deprecated "::set-env" command in the github CI runner - since 2 weeks or so.  A fix is already in #1622  (patch `github CI: Fix broken "::set-env". Use the new $GITHUB_ENV `) since some time. But @mhammond seems not to be around currently.

**Comment by tobias-loew** [comment]
I made a local branch with the setenv-fix: all tests (but one that seems to hang) passed (https://github.com/tobias-loew/pywin32/actions/runs/390146137)

File: com/win32com/client/build.py [hunk 1]
```diff
@@ -571,7 +645,7 @@ def MakeDefaultArgRepr(defArgVal):
     return repr(val)
   return None
 
-def BuildCallList(fdesc, names, defNamedOptArg, defNamedNotOptArg, defUnnamedArg, defOutArg, is_comment = False):
+def BuildCallList(fdesc, names, defNamedOptArg, defNamedNotOptArg, defUnnamedArg, defOutArg, is_comment = False, bTypeHints = 0):
```
Comment (mhammond) [inline com/win32com/client/build.py:648]: Please use `False` as the default.

**Review by mhammond — COMMENTED** [review]
I had a quick look as I'm very busy at the moment, so might take some time to get back to this. I can't even remember what's in the .zip file you attached previously, but I do note that this really needs some comments - eg, what exactly is an enum in this context? What is a type-hint? When would I want to enable those options? When would I *not* want to enable those options? How is the large `NativeTypeMap` object used, etc? Ideally we'd also have tests, maybe by abusing PyCOMTest.idl?

**Comment by mhammond** [comment]
Thanks for the contribution, and feel free to reopen this if you end up making the requested changes.

**Pull request closed without merging**